### PR TITLE
I vs and moves

### PR DESCRIPTION
**Pull request by sbets1028**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by MangoScango** [comment]
I mean I guess it might be worth posting a separate PR just to not clutter the other one, I would have prefered to post this myself though. Oh well.

This is based heavily on #332 . Uses the same DB schema, though only gathers IVs and moves. Sends the info over the webhook too. The args are --encounter, -- encounter-delay, and --encounter-blacklist.

**Comment by eugenio412** [comment]
Can't understand how to activate the encounter to true from the config.ini, just encounter: True doesen't work, using -enc works

**Comment by nhumber** [comment]
Not much else to say but.. Amazing!!!!

**Comment by MrGrimReaper** [comment]
Missing the colon on 
>#encounter-delay

**Comment by nhumber** [comment]
ok after using this for 12 hours or more, i can say there is some issues, i get a fair bit of 0,0,0 with this branch, i added more accounts thinking it needed more time before the next spawnpoint needed scanning, but thats not it, something in the code is creating 0,0,0 workers alot, i used the same accounts and went back to dev branch and everything is fine with the accounts i see no 0,0,0 

running -ss mode with 20 workers on 4800 spawn points, 4km sq area.

tried with -ed 1 , 2 , 0 all no changes.

**Comment by MangoScango** [comment]
That's unfortunate. I've seen no changes in my amount of 0/0/0, but my scan area is smaller. O think there might be an upper limit on the area you can scan from one ip, and this makes it much easier to hit that.

I want to write something to help debug where these are coming from, hopefully I get time to soon.

**Comment by nhumber** [comment]
if you do that i will debug it, 

its weird because i get 0,0,0 even before any encounter for IV is made, i have a big blacklist so its not like its overloading with IV check(mainly checking the standard heavy hitters, 2-5 IV checks an hour), its not an IP ban because i can run easily 26-30 accounts on this IP without issue, and a simple restart back on dev branch with same accounts within a minute doesnt show 0,0,0 with same accounts like you would expect with an IP ban. 

also when i watch the output the 0,0,0 tend to move around through all accounts and not just a few.

only thing i havent tried is running your IV branch without encounters enabled to see if i get the same 0,0,0 upserts. (should have already tried that)

**Comment by MangoScango** [comment]
Yeah, actually do try that. Because one things I changed in this was
stopping parsing and upserting any Pokemon whose encounter id is already in
the db,  but now I realize that means Pokemon who spawn multiple times per
hour won't get picked up after the first time. I know how I can fix this
when I get home. I'm not sure how common these points are, so I don't know
how much that would account for. But that could be it.

**Comment by nhumber** [comment]
ok thanks, ill do that and do some more testing, 

what is a good value for encounter delay?

if i had to guess it would be about 10% of scans show 0,0,0.

**Comment by MangoScango** [comment]
I use 2, but I think I can get away with 1. I'm not sure how much it
matters.

**Comment by azd2** [comment]
Oh, you opened a new Pr with your own code? Nice :-) 
Still works amazing, use it since yesterday. Absolutely no problems and some really happy friends in my telegram channel :D (see PR #1187 for details, this new PR is what i used for it because webhooks was included). @MangoScango thank you again :-)

**Comment by neutron666** [comment]
@MangoScango @azd2 Would it be possible to include the CP of the Pokemon into this whole thing ? If i understand it right, when an encounter starts, this info should be available too.

**Comment by azd2** [comment]
Do your scanner accounts have lvl 30+? If not, you wont see the cp the pokemon will have.
If you scan with lvl 1 accounts, they will always see the cp the pokemon has at lvl 1.
So: would surely be possible, but senseless...

**Comment by neutron666** [comment]
Yes, sorry, my fault, was not thinking about the player levels :P Just forget it.

File: pogom/utils.py [hunk 1]
```diff
@@ -285,7 +285,7 @@ def get_args():
             sys.exit(1)
 
         args.encounter_blacklist = [int(i) for i in args.encounter_blacklist]
-
```
Comment (slideup-benni) [inline pogom/utils.py]: :+1:

**Comment by MangoScango** [comment]
So this branch makes the following changes:

- 3 new options
 - --encounter (or -enc) will enable the map to start encounters to pull more info
 - --encounter-delay (or -ed) sets the time to wait after finding a pokemon before encountering it. Defaults to 1, mustn't be 0.
 - --**encounter-blacklist** (or -eblk) will let you set pokemon ids to not encounter.

Using encounter-blacklist is HIGHLY recommended. Every pokemon you encounter requires an entire new request, potentially DOUBLING the amount of requests your server makes. This can easily overwhelm large maps and lead to bans. You can set encounter-blacklist in the config.ini with a list like so: [10,13,16]. Setting it via cli is more annoying, you have to set each id like "-eblk 10 -eblk 13". If anyone know a simple way to use a list in the cli, that'd be great.

The database gets bumped to 8, and gains 'individual_attack', 'individual_defense', 'individual_stamina', 'move_1', and 'move_2' in the pokemon table. Where the IVs are just the ints you'd expect, and the moves are numbers we look up in the JSON file provided.

The map.js is updated to calculate and **show percent of perfection** each pokemon has, as well as its **IVs and moves**. It also features **notifications**, using #1187 's implementation.

The code is based heavily on #332.

A notable change is that if a scanned pokemon's encounter id is in the database, and the encounter is still active, this pokemon won't be upserted again. I did this avoid wiping stats of rescanned pokemon, because I don't make encounter requests for pokemon we've already encountered. 

To compare this branch to #1187, the important differences seems to be that I pass the api from search.py to parse_map, and do the work there. He does it in search.py. His branch will always make the encounter request even if it's for a pokemon we already know, mine doesn't. His branch will work retroactively and get IVs for old pokemon because of this though, mine won't.

I've had some reports of increased 0/0/0 responses. In my own experience, I haven't seen an increase, but my map is relatively small, apparently. My emptys seem to be related to the auth token being renewed.

SO! I think that about covers it. If you guys want me to submit this PR with this info where it actually belongs, I can.

**Comment by azd2** [comment]
Nicely summarized. Implementation of webhooks notification for new values was the reason i tried your PR in the 1st place :-) Don't like that this seems to evolve into some kind of race between you and Owraight though :-( I would have preferred you two working together to get this included somehow...
Well, if a decision has to be made, i hope they decide for this PR since (at least to me) it looks like this is ready to be included as it is.

**Comment by MangoScango** [comment]
@azd2 Yeah, it was never my intention to compete with anything anyone else was doing, this was just something that I was working on before he posted his PR, and I saw it had some advantages.

**Comment by sbets1028** [comment]
I am getting  horrible error, "ImportError: No module named cachetools"

**Comment by lota7** [comment]
Adding a "whitelist" instead of blacklist would be alot better.
better having 10 ids in that list than having 140.

**Comment by jorgebsa** [comment]
I'm of the same opinion. If this gets merged, it should include whitelisted and not blacklisted pokémon. Having to over 100 ids to a blacklist to every worker's config instead of just a handful to a whitelist is kinda stupid imo.

**Comment by Spiritvn** [comment]
Everyone have different needs @jorgebsa but i also vote for the white list (i'm using that way) because for me, the include list always shorter than exclude.

**Comment by hd1080** [comment]
Maybe the best would be to support a whitelist and a blacklist, of course not at the same time :)
For me a blacklist is better.

**Comment by MangoScango** [comment]
Whitelist is fine by me. I picked a blacklist because there were lots of
reports of people running stuff like this on everything without issue. So i
figured a blacklist would be sufficient. I'll either switch it or add
whitelist support later today.

**Comment by Spiritvn** [comment]
@MangoScango true i agree everyone has different targets, it's better to add an argument to know if it's whitelist/blacklist that people can switch it themself.

**Comment by Imoq** [comment]
Just please, for the sake of everybody (ejem: people who run the map without even looking at most options), please make whitelist the default option when run -iv , that way at least people will have to read how to enable some of the scanned Pokémon for the potential extra load to servers.

**Comment by MangoScango** [comment]
That's a good point. Will change it to default to whitelist, with a flag to
flip it to black.

**Comment by nhumber** [comment]
Me again still getting 0,0,0 responses. Not sure what's up but main branch doesn't do this. I can't be the only one..

**Comment by MangoScango** [comment]
@nhumber I can't seem to reproduce this. Are you running a beehive by chance? I've only throughly tested spawnpoint.

**Comment by azd2** [comment]
I am running a beehive. Getting empty scans sometimes too. Is that a problem? I thought a scanpoint gets 0,0,0 if there are no pokemon at the point and no changes to nearby stops/gyms. Is that wrong/does it mean empty spawnpoints are always a problem?
I can say that the number of found pokemon per 24 hours did not decrease since fully switching to this pr on friday.

**Comment by neutron666** [comment]
I'm running this on a linux box using CentOS 7, mysql, nothing special...running my map with spawnpoint scanning only and had ONE Pokemon with 0/0/0 in the last 24 hours. So spawnpoint scanning doesn't seem to be the problem.

**Comment by MangoScango** [comment]
Well I made a change in the way Pokemon are parsed so that Pokemon that you've already seen won't get parsed again, which would result in a 0 find for Pokemon if you rescan them. If this coincides with a legitimate 0/0 respone for gyms and pokestops, then this is expected behavior and nothing is actually wrong. I should update this to not report empty in this case though.

**Comment by nhumber** [comment]
@MangoScango ,using  -ss mode 4800 points, 20 accounts, multiple instances( a server only instance, a gym only instance, and my scanning instance(20 accounts doing pokemon only)) ill do some more testing here myself, ill build a spawn.json of just say 500m circle and run it for a bit and see if the 0,0,0 still occurs.

**Comment by MangoScango** [comment]
@nhumber A pokemon only map getting lots emptys is expected with a beehive, but with spawnpoint scanning I can't really explain it. That change is the only one I can think of that would cause this.

You're sure this doesn't happen with main branch? It going through all the accounts sounds like the auth token refresh issue.

**Comment by MangoScango** [comment]
Try the the -v arg and output to a file, then you can ctrl+f for "0 finds" we can get a better idea what's going on when this happens.

**Comment by nhumber** [comment]
ok, just tried on a fresh DB wipe and fresh download and install, getting this now.. seems 0,0,0 is gone but whatever this error is must be an issue.

2016-09-11 08:53:33,418 [search-worker-10][        models][    INFO] Parsing found 1 pokemons, 0 pokestops, and 0 gyms
/usr/local/lib/python2.7/dist-packages/pymysql/cursors.py:161: Warning: Truncated incorrect DOUBLE value: 'MjEzMzMxOTEyMjQ4NzE3ODAyOQ=='
  result = self._query(query)
/usr/local/lib/python2.7/dist-packages/pymysql/cursors.py:161: Warning: Truncated incorrect DOUBLE value: '2016-09-11 12:53:33.367659'

**Comment by MangoScango** [comment]
@nhumber yeah, that's probably an issue with the way I'm querying the database to look for previously encountered pokemon. My inexperience with databases is showing. Haha. Will look into that when I get home from work.

**Comment by nhumber** [comment]
@MangoScango ,Thanks again, curious as to why im only seeing it tho... ill mess around a bit and try a few things.

**Comment by justsomedudeonthenet** [comment]
So, everyone is pointing me here as this being the best version of the IV scanning.  As I said over there, I really want to help get this feature in.

One problem right off the bat - the devs aren't going to read all the comments.  They tend to ignore PRs that didn't bother to fill out the post.  So please update the PR description with all the information in the comments - what features it adds, what switches it uses, etc.

**Comment by MangoScango** [comment]
@justsomedudeonthenet I didn't start this PR, so I can't do that. Should a
start a new one then?

**Comment by justsomedudeonthenet** [comment]
Closing this PR because the developer should have access to change their own PR!

@MangoScango please start a new PR so you'll be able to edit it appropriately.

**Comment by nhumber** [comment]
@mangoscango, looks like whatever i had done yesterday i must have missed your commit somehow, currently all is working good, just getting that Warning: Truncated incorrect DOUBLE value: constantly, doesnt seem to be affecting the map runnung, been going for 6 hours now strong, Zero 0,0,0 to report!!
Thanks, good luck on the new pull and may the merge gods be good to u ;-)

**Comment by sLoPPydrive** [comment]
Just found something regarding performance. I wondered why my MySQL performed so bad with this PR and found the following query (and similar ones) which took about 600+ms on my machine:
```
SELECT `t1`.`encounter_id`, 
       `t1`.`spawnpoint_id`, 
       `t1`.`pokemon_id`, 
       `t1`.`latitude`, 
       `t1`.`longitude`, 
       `t1`.`disappear_time`, 
       `t1`.`individual_attack`, 
       `t1`.`individual_defense`, 
       `t1`.`individual_stamina`, 
       `t1`.`move_1`, 
       `t1`.`move_2` 
FROM   `pokemon` AS t1 
WHERE  ( ( 'NzUwNTI1MTE4NDQ0MzAzOTcyNQ==' 
           AND `t1`.`disappear_time` ) > '2016-09-11 19:01:48.829000' ) 
```
I didn't quite understand what should be selected here. There was no result at all. The matching line in code is most probably:
```
query = (Pokemon
    .select()
    .where(Pokemon.encounter_id == b64encode(str(encounter_id)) &
        Pokemon.disappear_time > datetime.utcnow())
    .dicts()
    )
```
I understand the code but the syntax seems to be wrong (I'm no python guy). However according to the peewee documentation I rewrote the statement as follows (more brackets!):
```
query = (Pokemon
    .select()
    .where((Pokemon.encounter_id == b64encode(str(encounter_id))) &
        (Pokemon.disappear_time > datetime.utcnow()))
    .dicts())
```
This made the slow queries disappear (and actually return a result!).

Again, I'm not sure that I did the right thing but my intuition tells me that there is indeed a bug in those lines.

**Comment by MangoScango** [comment]
@nhumber Great! Hopefully I can get whatever that error is sorted out too here in a minute.

@sLoPPydrive Doh! Yeah, you're definitely right. Thanks, I don't know if I would have caught that.

**Comment by Spiritvn** [comment]
I also have the Double/Interger value warning when try to use MySQL instead of SQLite.

**Comment by MangoScango** [comment]
I can't seem to replicate this MySQL error. I've been using a MySQL as my primary database for awhile and haven't notice it. So I'm gonna need some help squashing that.

**Comment by TheRoastedBean** [comment]
Changing the database schema from 5 ==> 8 was an exercise in frustration as well.  In order for this commit to work, had to start, stop the map, then downgrade the version since the extra columns in the pokemon table were never created.  Using mysql 5.7, If it helps, here is the exception being thrown.

```
376695 2016-09-11 16:06:25,295 [search-worker-0][        pgoapi][    INFO] Cleanup of request!
376696 2016-09-11 16:06:25,297 [search-worker-0][        peewee][   DEBUG] ('SELECT `t1`.`encounter_id`, `t1`.`spawnpoint_id`, `t1`.`pokemon_id`, `t1`.`latitude`, `t1`.`longitude`, `t1`.`disappear_time`, `t1`.`individual_attack`, `t1`.`individual_defense`, `t1`.`individual_stamina`, `t1`.`move_1`, `t1`.`move_2` FROM `pokemon` AS t1 WHERE ((%s AND `t1`.`disappear_time`) > %s)', ['NTkxMDQ4NjY2NjEzNzEyNzc=',datetime.datetime(2016, 9, 11, 20, 6, 25, 295941)])
376697 2016-09-11 16:06:25,305 [wh-updater-0][connectionpool][INFO] Starting new HTTP connection (1): 127.0.0.1
376698 2016-09-11 16:06:25,310 [wh-updater-0][connectionpool][DEBUG] "POST / HTTP/1.1" 200 2
376699 /home/[user]/venvs/ivmap/local/lib/python2.7/site-packages/pymysql/cursors.py:161: Warning: Truncated incorrect INTEGER value: 'NTkxMDQ4NjY2NjEzNzEyNzc='
376700   result = self._query(query)
376701 /home/[user]/venvs/ivmap/local/lib/python2.7/site-packages/pymysql/cursors.py:161: Warning: Truncated incorrect DOUBLE value: 'NTkxMDQ4NjY2NjEzNzEyNzc='
376702   result = self._query(query)
376703 /home/[user]/venvs/ivmap/local/lib/python2.7/site-packages/pymysql/cursors.py:161: Warning: Truncated incorrect DOUBLE value: '2016-09-11 20:06:25.295941'
376704   result = self._query(query)
```

**Comment by weizh888** [comment]
@eugenio412 Hi, I'm using your branch, and saw the IVs sent, but didn't see the moves. Are they included in your program, and how to enable it? Meanwhile, is that possible to send more pokemons that are with high IVs (like 90%), not limited to the defined list?
Thanks

**Comment by MangoScango** [comment]
@TheRoastedBean There's definitely some issues with database migration unrelated to this PR. Not sure if that's where this it's coming from. Needs looking into for sure.

**Comment by eugenio412** [comment]
@weizh888 working on that, there is a lot of work to do, still need time

**Comment by weizh888** [comment]
@eugenio412 Great to hear that! BTW, do you know how to change the location to scan for PokemonGoMap, except through defining the scanning steps?

**Comment by TheRoastedBean** [comment]
FYI the truncation warnings are due to a malformed query in the 
 `def get_encountered_pokemon(encounter_id)` method.  It turns out that @sLoPPydrive fixed it with his fix.  The following where clause was off
```
WHERE ((%s AND `t1`.`disappear_time`) > %s)'
```
and should have been
```
WHERE ((`t1`.`encounter_id` = %s)
```
Most likely due to the parentheses, thanks @sLoPPydrive for cleaning that up.  This is solved as far as I'm concerned.

**Comment by MangoScango** [comment]
Glad to hear. @sloppydrive is da real MVP.

**Comment by weizh888** [comment]
@eugenio412 I just saw your update on PogomBOT. But that ignored the detailed information of (attack, defense, stamina). Instead, it only gives an IV value. Could you refer to MangoScango and modify the code? Thanks.

**Comment by eugenio412** [comment]
What do you mean? @weizh888, we are using the IV formula, there is no work to do for @mangoscango

**Comment by weizh888** [comment]
@eugenio412 In the latest version of PogomBOT, there is one sentence "iv = pokemon.getIVs()". In that function, it gets the IV values ditectly (possibly from the database?). While in the previous version, it used the three features to calculate IV. It worked fine when I added one if sentence to filter low IV pokemons, but the updated one doesn't work for me.

**Comment by eugenio412** [comment]
We are going to add notifications based on Iv if that's your concern, today we had a huge commit, it's far more elegant in the code than the previous, even if it could be harder to modify @weizh888

**Comment by weizh888** [comment]
@eugenio412 Hope to see the updates! It is more difficult to modify. Thanks.

**Comment by weizh888** [comment]
@eugenio412 Is the PogomBot compatible with https://github.com/PokemonGoMap/PokemonGo-Map/pull/1272 ? BTW, is it possible to scan multiple pogom .db files and send the notifications? I have to run multiple terminals for scanning multiple PokemonGo-Map.

**Comment by n3cr0cr0w** [comment]
@weizh888 myself on one server, have them all connected to the same database.

**Comment by eugenio412** [comment]
@weizh888  yes it is, just added IV based notification in our develop branch

**Comment by danimorelo** [comment]
@TheRoastedBean thanks for pointing this out, manually downgrading to version 7 did the trick! so it went from 5->8->7->8

**Pull request closed without merging**